File: leetcode/Go/trees/112/112.go
package main

/*
*void Postorder(TreeNode * root, vector<int> &v)
    {
        if(root == nullptr)
            return;
        else
        {
            Postorder(root->left, v);
            Postorder(root->right, v);
            v.push_back(root->val);
        }
    }
*/
// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	hasSum := false
	// If root is null
	if root == nil {
		return false
	}
	traverseTree(root, 0, targetSum, &hasSum)
	return hasSum
}

func traverseTree(root *TreeNode, sum, targetSum int, hasSum *bool) {
	// In case we found a match
	if *hasSum {
		return
	}
	// Setting base recursion
	if root == nil {
		return
	}
	sum += root.Val
	// If we find a leaf node and match the target
	if sum == targetSum && root.Left == nil && root.Right == nil {
		*hasSum = true
	}
	traverseTree(root.Left, sum, targetSum, hasSum)
	traverseTree(root.Right, sum, targetSum, hasSum)
}

File: leetcode/Python/trees/112.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def traverse(root, sumPath):
            if root is None:
                return False
            sumPath += root.val
            # if there is a leaf check if is the target
            if root.left is None and root.right is None:
                return sumPath == targetSum
            return traverse(root.left, sumPath) or traverse(root.right, sumPath)
        return traverse(root, 0)

        
